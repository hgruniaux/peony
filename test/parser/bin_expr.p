//# RUN: peony %s

fn foo() -> f32 { return 0.0; }
fn bar() -> i32 { return 0; }

fn f() {
    5 + 8;
    2 + 3 * 7;
    1.0 / 7.0;
    3.0 - foo();

    true && false;
    true || false;

    5 && 2.0;
    //~^ ERROR: mismatched types; expected 'bool', found 'i32'
    //~| ERROR: mismatched types; expected 'bool', found 'f32'
    5 || 2.0;
    //~^ ERROR: mismatched types; expected 'bool', found 'i32'
    //~| ERROR: mismatched types; expected 'bool', found 'f32'

    if 5 == 1 {}
    if 8.0 >= 3.0 {}

    if 5 == 1.0 {}
    //~^ ERROR: cannot apply binary operator '==' to types 'i32' and 'f32'
    if 5 != 1.0 {}
    //~^ ERROR: cannot apply binary operator '!=' to types 'i32' and 'f32'
    if 5 < 1.0 {}
    //~^ ERROR: cannot apply binary operator '<' to types 'i32' and 'f32'
    if 5 <= 1.0 {}
    //~^ ERROR: cannot apply binary operator '<=' to types 'i32' and 'f32'
    if 5 > 1.0 {}
    //~^ ERROR: cannot apply binary operator '>' to types 'i32' and 'f32'
    if 5 >= 1.0 {}
    //~^ ERROR: cannot apply binary operator '>=' to types 'i32' and 'f32'

    5 & 8;
    5 ^ 8;
    5 | 8;
    5 << 8;
    5 >> 8;
    let a0 = 5;
    a0 &= 8;
    a0 ^= 8;
    a0 |= 8;
    a0 <<= 8;
    a0 >>= 8;

    true & 9.0;
    //~^ ERROR: mismatched types; expected '{integer}', found 'bool'
    //~| ERROR: mismatched types; expected '{integer}', found 'f32'
    true ^ 9.0;
    //~^ ERROR: mismatched types; expected '{integer}', found 'bool'
    //~| ERROR: mismatched types; expected '{integer}', found 'f32'
    true | 9.0;
    //~^ ERROR: mismatched types; expected '{integer}', found 'bool'
    //~| ERROR: mismatched types; expected '{integer}', found 'f32'
    true << 9.0;
    //~^ ERROR: mismatched types; expected '{integer}', found 'bool'
    //~| ERROR: mismatched types; expected '{integer}', found 'f32'
    true >> 9.0;
    //~^ ERROR: mismatched types; expected '{integer}', found 'bool'
    //~| ERROR: mismatched types; expected '{integer}', found 'f32'

    a0 ^= 0.0;
    //~^ ERROR: mismatched types; expected 'i32', found 'f32'

    let a1 = 0.0;
    a1 &= 5;
    //~^ ERROR: binary assignment operation '&=' cannot be applied to type 'f32'
    a1 ^= 5;
    //~^ ERROR: binary assignment operation '^=' cannot be applied to type 'f32'
    a1 |= 5;
    //~^ ERROR: binary assignment operation '|=' cannot be applied to type 'f32'
    a1 >>= 5;
    //~^ ERROR: binary assignment operation '>>=' cannot be applied to type 'f32'
    a1 <<= 5;
    //~^ ERROR: binary assignment operation '<<=' cannot be applied to type 'f32'

    // Special messages for classic operators.
    5 + true;
    //~^ ERROR: cannot add 'i32' to 'bool'
    a1 += true;
    //~^ ERROR: cannot add-assign 'bool' to 'f32'
    5u32 - 2f64;
    //~^ ERROR: cannot subtract 'f64' from 'u32'
    a1 -= true;
    //~^ ERROR: cannot subtract-assign 'bool' from 'f32'
    8.0f32 * 16f64;
    //~^ ERROR: cannot multiply 'f32' by 'f64'
    a1 *= true;
    //~^ ERROR: cannot multiply-assign 'f32' by 'bool'
    true / 5;
    //~^ ERROR: cannot divide 'bool' by 'i32'
    a1 /= true;
    //~^ ERROR: cannot divide-assign 'f32' by 'bool'
}
