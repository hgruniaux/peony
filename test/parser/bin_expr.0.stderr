<source>:15:7: error: mismatched types; expected 'bool', found 'i32'
   15 |     5 && 2.0;
      |     ~    ~~~
<source>:15:7: error: mismatched types; expected 'bool', found 'f32'
   15 |     5 && 2.0;
      |     ~    ~~~
<source>:18:7: error: mismatched types; expected 'bool', found 'i32'
   18 |     5 || 2.0;
      |     ~    ~~~
<source>:18:7: error: mismatched types; expected 'bool', found 'f32'
   18 |     5 || 2.0;
      |     ~    ~~~
<source>:25:10: error: cannot apply binary operator '==' to types 'i32' and 'f32'
   25 |     if 5 == 1.0 {}
      |        ~ ^  ~~~
<source>:27:10: error: cannot apply binary operator '!=' to types 'i32' and 'f32'
   27 |     if 5 != 1.0 {}
      |        ~ ^  ~~~
<source>:29:10: error: cannot apply binary operator '<' to types 'i32' and 'f32'
   29 |     if 5 < 1.0 {}
      |        ~ ^ ~~~
<source>:31:10: error: cannot apply binary operator '<=' to types 'i32' and 'f32'
   31 |     if 5 <= 1.0 {}
      |        ~ ^  ~~~
<source>:33:10: error: cannot apply binary operator '>' to types 'i32' and 'f32'
   33 |     if 5 > 1.0 {}
      |        ~ ^ ~~~
<source>:35:10: error: cannot apply binary operator '>=' to types 'i32' and 'f32'
   35 |     if 5 >= 1.0 {}
      |        ~ ^  ~~~
<source>:50:10: error: mismatched types; expected '{integer}', found 'bool'
   50 |     true & 9.0;
      |     ~~~~
<source>:50:10: error: mismatched types; expected '{integer}', found 'f32'
   50 |     true & 9.0;
      |            ~~~
<source>:53:10: error: mismatched types; expected '{integer}', found 'bool'
   53 |     true ^ 9.0;
      |     ~~~~
<source>:53:10: error: mismatched types; expected '{integer}', found 'f32'
   53 |     true ^ 9.0;
      |            ~~~
<source>:56:10: error: mismatched types; expected '{integer}', found 'bool'
   56 |     true | 9.0;
      |     ~~~~
<source>:56:10: error: mismatched types; expected '{integer}', found 'f32'
   56 |     true | 9.0;
      |            ~~~
<source>:59:10: error: mismatched types; expected '{integer}', found 'bool'
   59 |     true << 9.0;
      |     ~~~~
<source>:59:10: error: mismatched types; expected '{integer}', found 'f32'
   59 |     true << 9.0;
      |             ~~~
<source>:62:10: error: mismatched types; expected '{integer}', found 'bool'
   62 |     true >> 9.0;
      |     ~~~~
<source>:62:10: error: mismatched types; expected '{integer}', found 'f32'
   62 |     true >> 9.0;
      |             ~~~
<source>:66:8: error: mismatched types; expected 'i32', found 'f32'
   66 |     a0 ^= 0.0;
      |           ~~~
<source>:70:8: error: binary assignment operation '&=' cannot be applied to type 'f32'
   70 |     a1 &= 5;
      |        ^
<source>:72:8: error: binary assignment operation '^=' cannot be applied to type 'f32'
   72 |     a1 ^= 5;
      |        ^
<source>:74:8: error: binary assignment operation '|=' cannot be applied to type 'f32'
   74 |     a1 |= 5;
      |        ^
<source>:76:8: error: binary assignment operation '>>=' cannot be applied to type 'f32'
   76 |     a1 >>= 5;
      |        ^
<source>:78:8: error: binary assignment operation '<<=' cannot be applied to type 'f32'
   78 |     a1 <<= 5;
      |        ^
<source>:82:7: error: cannot add 'i32' to 'bool'
   82 |     5 + true;
      |     ~ ^ ~~~~
<source>:84:8: error: cannot add-assign 'bool' to 'f32'
   84 |     a1 += true;
      |     ~~ ^  ~~~~
<source>:86:10: error: cannot subtract 'f64' from 'u32'
   86 |     5u32 - 2f64;
      |     ~~~~ ^ ~~~~
<source>:88:8: error: cannot subtract-assign 'bool' from 'f32'
   88 |     a1 -= true;
      |     ~~ ^  ~~~~
<source>:90:12: error: cannot multiply 'f32' by 'f64'
   90 |     8.0f32 * 16f64;
      |     ~~~~~~ ^ ~~~~~
<source>:92:8: error: cannot multiply-assign 'f32' by 'bool'
   92 |     a1 *= true;
      |     ~~ ^  ~~~~
<source>:94:10: error: cannot divide 'bool' by 'i32'
   94 |     true / 5;
      |     ~~~~ ^ ~
<source>:96:8: error: cannot divide-assign 'f32' by 'bool'
   96 |     a1 /= true;
      |     ~~ ^  ~~~~
